application/src/mmul.cu

```diff
@@ -1,20 +1,43 @@
 #include <mmul.cuh>
 
-__global__ void mmul_gpu(double *a, double *b, double *c, int m, int n, int k) {
-  int row = blockIdx.y * blockDim.y + threadIdx.y;
-  int col = blockIdx.x * blockDim.x + threadIdx.x;
-  double sum = 0;
-  if (col < k && row < m) {
-    for (int i = 0; i < n; i++) {
-      sum += a[row * n + i] * b[i * k + col];
+void mmul_gpu(cuda_executor<oneway_t, bulk_t, blocking_t::always_t> &ex,
+              MatrixXd &a, MatrixXd &b, MatrixXd &c) {
+  double *a_d, *b_d, *c_d;
+
+  auto device_upload = [=](void *var, std::size_t size) {
+    void *gpu_var;
+    cudaMallocManaged(&gpu_var, size);
+    memcpy(gpu_var, var, size);
+    return gpu_var;
+  };
```

Comment:
> defined somewhere outside this scope as a normal function for reuse purposes
Since this only for testing purposes and there is currently no reuse so used a lambda, maybe later can separate it.

> taking a reference to a generic type and infer it's size automatically
Not sure how this is supposed to work

Reply:
```cpp
template<typename T>
void* device_upload(const T& var)
{
    void *gpu_var;
    cudaMallocManaged(&gpu_var, sizeof(var));
    std::memcpy(gpu_var, &var, sizeof(var));
    return gpu_var;
}
```

Reply:
This won't work if you are passing pointers, to copy. I guess it would be nice to provide an overload maybe.

Reply:
overload?
```cpp
template<typename T>
void* device_upload(const T* var)
{
    void *gpu_var;
    cudaMallocManaged(&gpu_var, sizeof(T));
    std::memcpy(gpu_var, var, sizeof(T));
    return gpu_var;
}
```

Reply:
This won't work for pointers right. It will just return the size of the type T, not the size of the memory pointed by the pointer.